File: linked_list/linkedlist.py
class Node:
    def __init__(self, val) -> None:
        self.val = val
        self.prev = None
        self.next = None
    
    def __str__(self):
        prev = self.prev.val if self.prev else None
        next = self.next.val if self.next else None
        return f"value: {self.val}, prev: {prev}, next: {next}"

a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')

a.next = b
b.prev = a
b.next = c
c.prev = b
c.next = d
d.prev = c

def iterate(root):
    current = root
    while current != None:
        print("Current: ", current.val)
        if current.prev != None:
            print("Previous: ", current.prev.val)
        if current.next != None:
            print("Next: ", current.next.val)
        current = current.next

def getHeadNode(root):
    head = root
    while head.prev != None:
        head = head.prev
    
    return head.val

def getTailNode(root):
    tail = root
    while tail.next != None:
        tail = tail.next
    
    return tail.val

def insertNode(node, parent):
    if parent.prev == None:
        parent.prev = node
        node.next = parent
    elif parent.next == None:
        parent.next = node
        node.prev = parent
    else:
        nextNode = parent.next
        
        parent.next = node
        nextNode.prev = node
        node.prev = parent
        node.next = nextNode

    return node.val

def deleteNode(node):
    if node.prev == None:
        node.next.prev = None
        node.next = None
    elif node.next == None:
        node.prev.next = None
        node.prev = None
    else:
        prevNode = node.prev
        nextNode = node.next

        prevNode.next = nextNode
        nextNode.prev = prevNode
        node.prev = None
        node.next = None

    return node.val

def deleteNodeAt(root, index):
    current = root
    count = 0
    while count != index:
        current = current.next
        print("current: ", count, current)
        if current == None:
            print("We are out of bounds of the array and the element was not found.")
            return False
        count += 1
    
    deleteNode(current)
    return current.val

def getLength(root):
    current = root
    results = []
    while current != None:
        results.append(current.val)
        current = current.next

    return len(results)


def getNode(root, targetVal):
    current = root
    while current.val != targetVal:
        current = current.next
        print('print', current)
        if current == None:
            return False
    
    return True

def sumOfLinkedList(root):
    totalSum = 0
    current = root
    while current != None:
        totalSum += ord(current.val)
        current = current.next
    
    return totalSum
# iterate(a)
# print("Head: ", getHeadNode(a))
# print("Tails: ", getTailNode(a))
# print("Insert: ", insertNode(Node(1), a))
# # print("Delete: ", deleteNode(b))
# print("Delete Node At: ", deleteNodeAt(a, 2))
# iterate(a)
# print("Get Node (Found): ", getNode(a, 'd'))
# print("Get Node (Not Found): ", getNode(a, 'f'))
# print("Length: ", getLength(a))
# one = Node(1)
# print("Insert head: ", insertNode(one, a))
# iterate(one)
# print("Delete head: ", deleteNode(one))
# iterate(a)
# print("Insert tail: ", insertNode(one, d))
# iterate(a)
# print("Delete tail: ", deleteNode(one))
# iterate(a)
print(sumOfLinkedList(a))

File: linked_list/linkedlist.js
class Node {
    constructor(val) {
        this.val = val;
        this.prev = null;
        this.next = null;
    }
}

const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');


a.next = b;
b.prev = a;
b.next = c;
c.prev = b;
c.next = d;
d.prev = c;

function traversalLinkedList(root) {
    let node = root;
    while(node) {
        console.log(node.val);
        node = node.next;
    }
}

function getHeadNode(node) {
    let head = node;
    while(head.prev) {
        head = head.prev;
    }

    return head.val;
}

function getTailNode(node) {
    let tail = node;
    while(tail.next) {
        tail = tail.next;
    }

    return tail.val;
}

function insertNode(node, parent) {
    if(!parent.prev) {
        parent.prev = node;
        node.next = parent;
    } else if (!parent.next) {
        parent.next = node;
        node.prev = parent;
    } else {
        let newChildNode = parent.next;

        parent.next = node;
        node.prev = parent;
        node.next = newChildNode;
        newChildNode.prev = node;
    }

    return node;
}

function deleteNode(node) {
    if(!node.prev) {
        node.next.prev = null;
        node.next = null;
    } else if(!node.next) {
        node.prev.next = null;
        node.prev = null;
    } else {
        let nodePrev = node.prev;
        let nodeNext = node.next;
    
        node.prev.next = nodeNext;
        node.next.prev = nodePrev;
        node.prev = null;
        node.next = null;    
    }

    return node;
}

function deleteAt(root, index) {
    let current = root;
    let count = 0;
    while(current) {
        if(count === index) {
            return deleteNode(current);
        }
        current = current.next;
        count++;
    }

    return -1;
}

function getNode(root, target) {
    let current = root;
    while(current){
        if(current.val === target){
            return current.val;
        } else {
            current = current.next;
        }  
    }
    
    return -1; 
}

function getLength(root) {
    let results = [];
    let current = root;
    while(current) {
        results.push(current.val);
        current = current.next;
    }
    console.log(results)
    return results.length;
}

function sumOfLinkedList(node) {
    let sum = 0;
    let current = node;

    while(current) {
        sum += current.val.charCodeAt(0);
        current = current.next;
    }

    return sum;
}

// traversalLinkedList(a);
// console.log("Head: ", getHeadNode(c));
// console.log("Tail: ", getTailNode(d));
// console.log("Insert: ", insertNode(new Node(1), a));
// console.log("Delete: ", deleteNode(b));
// traversalLinkedList(a);
// console.log("Insert: ", insertNode(b, a));
// traversalLinkedList(a);
// console.log("Get node: ", getNode(a, 'c'));
// console.log("Get node (Not found): ", getNode(a, 'v'));
// console.log("Delete node at: ", deleteAt(a, 2));
// console.log("Get size: ", getLength(a));
// let one = new Node(1);
// console.log("Insert to head: ", insertNode(one, a));
// traversalLinkedList(one);
// console.log("Delete from head: ", deleteNode(one))
// traversalLinkedList(one);
// console.log("Insert to tail: ", insertNode(one, d));
// traversalLinkedList(a);
// console.log("Delete from tail: ", deleteNode(one))
// traversalLinkedList(a);
console.log(sumOfLinkedList(a))

